src/script/lua_api/l_particleparams.h

```diff
@@ -0,0 +1,199 @@
+		if (lua_isnumber(L, -1)) { // shortcut for uniform vectors
+			auto n = lua_tonumber(L, -1);
+			ret = v2fParameter(n,n);
+		} else {
+			ret = (v2fParameter)check_v2f(L, -1);
+		}
+	}
+
+	inline void readLuaValue(lua_State* L, TweenStyle& ret) {
+		const char* const opts[] = { "fwd", "rev", "pulse", "flicker", NULL };
+		const TweenStyle optmap[] = {
+			TweenStyle::fwd,
+			TweenStyle::rev,
+			TweenStyle::pulse,
+			TweenStyle::flicker,
+		};
+		static_assert(
+				(sizeof opts / sizeof opts[0])-1 ==
+				(sizeof optmap / sizeof optmap[0]),
+				"option maps not synced");
+
+		if (lua_isstring(L, -1)) {
+			size_t v = luaL_checkoption(L, -1, opts[0], opts);
+			ret = optmap[v];
+		}
+	}
+
+	template <typename T> void
+	readLuaValue(lua_State* L, RangedParameter<T>& field) {
+		if (lua_isnil(L,-1))
+			return;
+		if (!lua_istable(L,-1)) // is this is just a literal value?
+			goto set_uniform;
+
+		lua_getfield(L, -1, "min");
+		// handle convenience syntax for non-range values
+		if (lua_isnil(L,-1)) {
+			lua_pop(L, 1);
+			goto set_uniform;
+		}
+		readLuaValue(L,field.min);
+		lua_pop(L, 1);
+
+		lua_getfield(L, -1, "max");
+		readLuaValue(L,field.max);
+		lua_pop(L, 1);
+
+		lua_getfield(L, -1, "bias");
+		if (!lua_isnil(L,-1))
+			readLuaValue(L,field.bias);
+		lua_pop(L, 1);
+		return;
+
+		set_uniform:
+			readLuaValue(L, field.min);
+			readLuaValue(L, field.max);
+	}
+
+	template <typename T> void
+	readLegacyValue(lua_State* L, const char* name, T& field) {}
+
+	template <typename T> void
+	readLegacyValue(lua_State* L, const char* name, RangedParameter<T>& field) {
+		int tbl = lua_gettop(L);
+		lua_pushliteral(L, "min");
+		lua_pushstring(L, name);
+		lua_concat(L, 2);
+		lua_gettable(L, tbl);
+		if (!lua_isnil(L, -1)) {
+			readLuaValue(L, field.min);
+		}
+		lua_settop(L, tbl);
+
+		lua_pushliteral(L, "max");
+		lua_pushstring(L, name);
+		lua_concat(L, 2);
+		lua_gettable(L, tbl);
+		if (!lua_isnil(L, -1)) {
+			readLuaValue(L, field.max);
+		}
+		lua_settop(L, tbl);
+	}
+
+
+	template <typename T> void
+	readTweenTable(lua_State* L, const char* name, TweenedParameter<T>& field)
+	{
+		int tbl = lua_gettop(L);
+		lua_pushstring(L, name);
+		lua_pushliteral(L, "_tween");
+		lua_concat(L, 2);
+		lua_gettable(L, tbl);
+		if(lua_istable(L, -1)) {
+			// get the starting value
+			lua_pushinteger(L, 1), lua_gettable(L, -2);
+			readLuaValue(L, field.start);
+			lua_pop(L, 1);
+
+			// get the final value -- use len instead of 2 so that this
+			// gracefully degrades if keyframe support is later added
+			lua_pushinteger(L, (lua_Integer)lua_objlen(L, -1)), lua_gettable(L, -2);
+			readLuaValue(L, field.end);
+			lua_pop(L, 1);
+
+			// get the effect settings
+			lua_pushliteral(L, "style"), lua_gettable(L, -2);
+			readLuaValue(L, field.style);
+			lua_pop(L, 1);
+
+			lua_pushliteral(L, "reps"), lua_gettable(L, -2);
+			readLuaValue(L, field.reps);
+			lua_pop(L, 1);
+
+			lua_pushliteral(L, "start"), lua_gettable(L, -2);
```

Comment:
simpler: https://pgl.yoyo.org/luai/i/lua_getfield

Reply:
agh right of course